### ISSUE TITLE
Alignment of loads/stores with constant offset ramps is problematic

### ISSUE THREAD

**Issue by dsharletg** [halide/Halide#3842]
I am seeing some code that is very inefficient because we don't understand the alignment of loads/stores with (aligned) constant ramps. For example, the following program:

    ImageParam in(UInt(16), 3);

    Var x("x"), y("y"), c("c");

    Func blur_y("blur_y");
    const int rate = 4;
    RDom ry(0, rate);
    blur_y(x, y, c) += cast<int32_t>(in(x, y + ry, c)) * 2;

    Func downsample_y("downsample_y");
    downsample_y(x, y, c) = cast<int16_t>(blur_y(x, y * rate, c));

    downsample_y
        .compute_root()
        .hexagon()
        .reorder(c, x, y)
        .vectorize(x, 64).bound(c, 0, 3);

    blur_y.compute_at(downsample_y, x)
        .vectorize(x)
        .update()
        .vectorize(x)
        .reorder(c, ry)
        .unroll(c);

    downsample_y.compile_to_lowered_stmt("/dev/stdout", blur_y.infer_arguments(), Text, get_host_target().with_feature(Target::HVX_128));

This produces:

```
for (downsample_y.s0.y, downsample_y.min.1, downsample_y.extent.1) {
  for (downsample_y.s0.x.x, 0, t12) {
    let downsample_y.s0.x.v24.base.s = min((downsample_y.s0.x.x*64), t13)
    allocate blur_y[int32 * 64 * 1 * 3]
    produce blur_y {
      for (blur_y.s0.c, 0, 3) {
        blur_y[ramp((blur_y.s0.c*64), 1, 64) aligned(64, 0)] = x64(0)
      }
      for (blur_y.s1.ry$x, 0, 4) {
        blur_y[ramp(0, 1, 64) aligned(0, 0)] = halide.hexagon.interleave.vw(halide.hexagon.add_mpy.vw.vh.vuh(halide.hexagon.deinterleave.vw(blur_y[ramp(0, 1, 64) aligned(0, 0)]), x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t14)), 1, 64)]))
        blur_y[ramp(64, 1, 64) aligned(0, 64)] = halide.hexagon.interleave.vw(halide.hexagon.add_mpy.vw.vh.vuh(halide.hexagon.deinterleave.vw(blur_y[ramp(64, 1, 64) aligned(0, 64)]), x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t15)), 1, 64)]))
        blur_y[ramp(128, 1, 64) aligned(0, 128)] = halide.hexagon.interleave.vw(halide.hexagon.add_mpy.vw.vh.vuh(halide.hexagon.deinterleave.vw(blur_y[ramp(128, 1, 64) aligned(0, 128)]), x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t16)), 1, 64)]))
      }
    }
    consume blur_y {
      for (downsample_y.s0.c, 0, 3) {
        downsample_y[ramp(((downsample_y.s0.c*downsample_y.stride.2) + (((downsample_y.s0.y*downsample_y.stride.1) + t17) + downsample_y.s0.x.v24.base.s)), 1, 64)] = halide.hexagon.pack.vw(blur_y[ramp((downsample_y.s0.c*64), 1, 64) aligned(64, 0)])
      }
    }
    free blur_y
  }
}
```

This code is problematic because the interleaves and deinterleaves should be removed, and add a lot of overhead. The reason these are not removed is that the EliminateInterleaves pass checks if the loads/stores are aligned. If I make this change:

```
diff --git a/src/HexagonOptimize.cpp b/src/HexagonOptimize.cpp
index a830bca00..b03a20367 100644
--- a/src/HexagonOptimize.cpp
+++ b/src/HexagonOptimize.cpp
@@ -1596,11 +1596,11 @@ class EliminateInterleaves : public IRMutator {
                 // buffer state.
             }
             internal_assert(aligned_buffer_access.contains(op->name) && "Buffer not found in scope");
-            bool &aligned_accesses = aligned_buffer_access.ref(op->name);
+            //bool &aligned_accesses = aligned_buffer_access.ref(op->name);
             int64_t aligned_offset = 0;
 
             if (!alignment_analyzer.is_aligned(op, &aligned_offset)) {
-                aligned_accesses = false;
+              //aligned_accesses = false;
             }
         }
         if (deinterleave_buffers.contains(op->name)) {
@@ -1630,11 +1630,11 @@ class EliminateInterleaves : public IRMutator {
                 // actually interleaved (and don't just yield an
                 // interleave).
                 internal_assert(aligned_buffer_access.contains(op->name) && "Buffer not found in scope");
-                bool &aligned_accesses = aligned_buffer_access.ref(op->name);
+                //bool &aligned_accesses = aligned_buffer_access.ref(op->name);
                 int64_t aligned_offset = 0;
 
                 if (!alignment_analyzer.is_aligned(op, &aligned_offset)) {
-                    aligned_accesses = false;
+                  //aligned_accesses = false;
                 }
```

The generated IR looks much better:
```
    produce blur_y {
      for (blur_y.s0.c, 0, 3) {
        blur_y[ramp((blur_y.s0.c*64), 1, 64) aligned(64, 0)] = x64(0)
      }
      for (blur_y.s1.ry$x, 0, 4) {
        blur_y[ramp(0, 1, 64) aligned(0, 0)] = halide.hexagon.add_mpy.vw.vh.vuh(blur_y[ramp(0, 1, 64) aligned(0, 0)], x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t14)), 1, 64)])
        blur_y[ramp(64, 1, 64) aligned(0, 64)] = halide.hexagon.add_mpy.vw.vh.vuh(blur_y[ramp(64, 1, 64) aligned(0, 64)], x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t15)), 1, 64)])
        blur_y[ramp(128, 1, 64) aligned(0, 128)] = halide.hexagon.add_mpy.vw.vh.vuh(blur_y[ramp(128, 1, 64) aligned(0, 128)], x64((int16)2), in[ramp(((((downsample_y.s0.y*4) + blur_y.s1.ry$x)*in.stride.1) + (downsample_y.s0.x.v24.base.s + t16)), 1, 64)])
      }
    }
    consume blur_y {
      for (downsample_y.s0.c, 0, 3) {
        downsample_y[ramp(((downsample_y.s0.c*downsample_y.stride.2) + (((downsample_y.s0.y*downsample_y.stride.1) + t17) + downsample_y.s0.x.v24.base.s)), 1, 64)] = halide.hexagon.trunc.vw(blur_y[ramp((downsample_y.s0.c*64), 1, 64) aligned(64, 0)])
      }
    }
```

**Comment by dsharletg** [comment]
This is with https://github.com/halide/Halide/pull/3841

**Comment by abadams** [comment]
These are maximally aligned, and correctly tagged as such. Does the alignment analyzer not understand that everything divides zero (the modulus here)?

**Comment by dsharletg** [comment]
It looks like some of the alignment information is being lost during Hexagon lowering:

Aligning loads for HVX....
blur_y 0 0
blur_y 0 0
blur_y 0 32
in 1 0
blur_y 0 64
blur_y 0 64
blur_y 0 96
in 1 0
blur_y 0 128
blur_y 0 128
blur_y 0 160
in 1 0
blur_y 64 0
blur_y 64 0
blur_y 64 32
Carrying values across loop iterations...
Eliminating boolean vectors from Hexagon code...
Optimizing Hexagon instructions...
blur_y 64 0
blur_y 1 0
blur_y 0 0
blur_y 1 0
blur_y 0 64
blur_y 1 0
blur_y 0 128
blur_y 1 0

This is the modulus and remainder for loads from the buffer named. They start out aligned, and then something loses that alignment, but only sometimes...

**Comment by dsharletg** [comment]
@pranavb-ca would this be something you would be able to take a look at?

**Comment by dsharletg** [comment]
I think I found the fix (#3844) while debating #3841 :)
